### ISSUE TITLE
Incorrectly allows trailing comments after section headers

### ISSUE THREAD

**Issue by tfausak** [haskell/cabal#9677]
**Describe the bug**
Typically Cabal does not allow trailing comments. All comments must be on their own line. They cannot follow a field name or value. However if you put an trailing comment after a section header, it is accepted by Cabal. 

**To Reproduce**
Create a package description like this:

``` cabal
cabal-version: >= 1.2
name: example
version: 0
build-type: Simple
executable example -- foo
  main-is: Main.hs
  build-depends: base
```

The only important line is `executable example -- foo`. Using `executable` specifically isn't necessary. Any section (like `library`) behaves the same. I used `executable` to show that the comment isn't being interpreted as an argument to the section. In other words, we're not defining an executable called `"example -- foo"`.

Similarly the `cabal-version` does not appear to be important. Setting it to, say, `cabal-version: 3.8` exhibits the same behavior.

To let Cabal build this example package successfully, also create a `Main.hs` file. It's not important what's in it. Something like this will work:

``` hs
main = pure ()
```

Finally, run `cabal build`.

**Expected behavior**
The build should fail due to invalid syntax in the package description. For example, consider the following package description:

``` cabal
name: example -- foo
```

That fails with the following error:

```
Errors encountered when parsing cabal file ./example.cabal:

example.cabal:1:15: error:
unexpected '-'
expecting space or end of input

    1 | name: example -- foo
      |               ^
```

**System information**
```
$ cabal --version
cabal-install version 3.10.2.1
compiled using version 3.10.2.1 of the Cabal library
```

**Additional context**
It seems like either trailing comments should be allowed everywhere or prohibited everywhere. I think the intent is for them to be prohibited. Various pieces of documentation and code say that:

https://github.com/haskell/cabal/blob/46a7bfc922d62477cc9de061a6dc0d3c91cdd9ee/doc/how-to-package-haskell-code.rst#L140-L142

https://github.com/haskell/cabal/blob/46a7bfc922d62477cc9de061a6dc0d3c91cdd9ee/Cabal-syntax/src/Distribution/Parsec.hs#L276-L278

**Comment by ulysses4ever** [comment]
Good catch! I wonder how many packages use this loophole... Certainly feels like an overlook but can break people if fixed.

**Comment by phadej** [comment]
The support for trailing comments is there for backward compatibility

E.g. https://hackage.haskell.org/package/hint-0.3.0.0/revision/0.cabal has 

```
  else { -- ghc < 6.8
      build-depends:    utf8-string < 0.3
  }
```

You can try to add 

```diff
diff --git a/Cabal-syntax/src/Distribution/Fields/Lexer.x b/Cabal-syntax/src/Distribution/Fields/Lexer.x
index b9b8ad54c..d4e637aba 100644
--- a/Cabal-syntax/src/Distribution/Fields/Lexer.x
+++ b/Cabal-syntax/src/Distribution/Fields/Lexer.x
@@ -107,7 +107,7 @@ tokens :-
 <in_section> {
   $spacetab+   ; --TODO: don't allow tab as leading space

-  "--" $comment* ;
+  "--" $comment*; --  { \pos len inp -> addWarningAt pos LexWarningComment >> lexToken }

   @name        { toki TokSym }
   @string      { \pos len inp -> return $! L pos (TokStr (B.take (len - 2) (B.tail inp))) }
diff --git a/Cabal-syntax/src/Distribution/Fields/LexerMonad.hs b/Cabal-syntax/src/Distribution/Fields/LexerMonad.hs
index 2aa53d2b5..12fa8d821 100644
--- a/Cabal-syntax/src/Distribution/Fields/LexerMonad.hs
+++ b/Cabal-syntax/src/Distribution/Fields/LexerMonad.hs
@@ -71,6 +71,8 @@ data LexWarningType
     LexInconsistentIndentation
   | -- | Brace syntax used
     LexBraces
+
+  | LexWarningComment
   deriving (Eq, Ord, Show)
```

and so on and check with `hackage-tests` how much "wrong" files there is. Might be a good idea to add a warning though, even if it cannot be reasonably made an error. (It could be made an error if there are not too many offending files to be fixed by hand using `Quirks` machinery).

**Comment by phadej** [comment]
I think it was bad design to lex section headers into tokens to begin with. I didn't know to question that (@dcoutts's?) decision back then. For field contents lexer doesn't do anything, it just munches the contents; I cannot come with a good reason to do differently for section headers. Cabal-syntax has to be able to parse conditionals from just a ByteString as that's needed elsewhere, so the extra lexing done by lexer is in a way wasted work leading to code duplication.
